includes/popups-analytics/class-popups-analytics-utils.php

```diff
@@ -0,0 +1,281 @@
+<?php
+/**
+ * Popups Analytics utils.
+ *
+ * @package Newspack
+ */
+
+use Google\Site_Kit\Modules\Analytics;
+use Google\Site_Kit\Context;
+use Google\Site_Kit_Dependencies\Google_Service_AnalyticsReporting_DateRange;
+use Google\Site_Kit_Dependencies\Google_Service_AnalyticsReporting_Metric;
+use Google\Site_Kit_Dependencies\Google_Service_AnalyticsReporting_ReportRequest;
+use Google\Site_Kit_Dependencies\Google_Service_AnalyticsReporting_Dimension;
+use Google\Site_Kit_Dependencies\Google_Service_AnalyticsReporting_GetReportsRequest;
+use Google\Site_Kit_Dependencies\Google_Service_AnalyticsReporting;
+use Google\Site_Kit_Dependencies\Google_Service_AnalyticsReporting_SegmentDimensionFilter;
+use Google\Site_Kit_Dependencies\Google_Service_AnalyticsReporting_DimensionFilterClause;
+use Google\Site_Kit\Core\Storage\Options;
+use Google\Site_Kit\Core\Storage\User_Options;
+use Google\Site_Kit\Core\Authentication\Authentication;
+
+/**
+ * Popup Analytics Utilities.
+ */
+class Popups_Analytics_Utils {
+	// Category is fixed, it identifies the GA custom event.
+	const EVENT_CATEGORY = 'Newspack Announcement';
+
+	/**
+	 * Get dates in range.
+	 *
+	 * @param Date   $start start date.
+	 * @param Date   $end end date.
+	 * @param string $format date format.
+	 */
+	private static function get_dates_in_range( $start, $end, $format = 'Y-m-d' ) {
+		return array_map(
+			function( $timestamp ) use ( $format ) {
+				return gmdate( $format, $timestamp );
+			},
+			range( strtotime( $start ) + ( $start < $end ? 4000 : 8000 ), strtotime( $end ) + ( $start < $end ? 8000 : 4000 ), 86400 )
+		);
+	}
+
+	/**
+	 * Date before.
+	 *
+	 * @param string $offset date offset in days.
+	 */
+	private static function date_before( $offset ) {
+		$today = new \DateTime();
+		$today = $today->modify( '-' . $offset . ' days' );
+		return $today->format( 'Y-m-d' );
+	}
+
+	/**
+	 * Fill in dates.
+	 *
+	 * @param array  $input array of days.
+	 * @param string $offset date offset in days.
+	 */
```

Comment:
Missing @return notation...

Reply:
Done in a6c2d0e3cced3d475964313d6282eaacf58d486c